### PR TITLE
Document the enum changes in RFC 2195

### LINKED ISSUE #244: Replace remaining "C-like enumerations" with "fieldless enumerations"
See rust-lang/rust#46348. Looks like a few snuck through: https://doc.rust-lang.org/nightly/reference/type-layout.html#reprc-enums, for instance.

### LINKED ISSUE #786: enum documentation not updated for RFC 2195
https://doc.rust-lang.org/reference/items/enumerations.html#custom-discriminant-values-for-field-less-enumerations talks about "Custom Discriminant Values for Field-Less Enumerations", but does not mention at all that since https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md Rust also supports custom discriminant values and defined layout for enums with fields.

### PR DESCRIPTION
**Pull request by poliorcetics**
Document the changes introduced in https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md.

Fixes #244 (I believe), fixes #786.

This is your RFC so I'll tag you @Gankra.

This is a restart from #246, applying the comments that were given at the time.

Disclaimer: I have never personally used this feature and only tested using the playground so I may be missing things, I'm sorry if that's the case.

### CODE REVIEW

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:376]: This sentence makes it sound like the discussion should be in
```
#### Primitive Representation of Enums With Fields
```
That subsection really has two parts: with and without `repr(C)`.

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:307]:
```suggestion
> Note: Due to the representation of `repr(C)` structs and unions, if a variant
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:311]:
```suggestion
> convenient or consistent for them.
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:375]: I think this span can be removed if the link has been removed.

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:391]:
```suggestion
For [field-less enums], primitive representations set the size and alignment to be the same as
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:397]:
```suggestion
The representation of a primitive representation enum is a `repr(C)` union of `repr(C)`
```

`repr(int)` isn't ever defined or used elsewhere. I realize the RFC used it as a shorthand for "primitive representation", and "primitive representation" is wordy, but I think it can be confusing when seen in isolation.

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:399]:
```suggestion
union is the primitive representation version of the enum with all fields removed ("the tag")
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:402]:
```suggestion
> Note: This representation is unchanged if the tag is given its own member in
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:404]:
```suggestion
> the union, should that make manipulation more clear for you (although to
> follow the C++ standard the tag member should be wrapped in
```

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md]: Good catch, I had completely forgotten it

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md]: I moved the paragraph to the end of the `Primitive Representation of Enums With Fields`, is it better ?

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:444]: Yes, much better. :) Some wording suggestions:
```suggestion
For enums with fields, it is also possible to combining `repr(C)` and a primitive representation.
This modifies the `repr(C)` by changing the
```
Also it would be nice to make "the `repr(C)`" in the 2nd line a link to the relevant section.

And finally, is it possible to summarize the difference between this and "normal" primitive reps in a sentence? Something like "union-of-structs vs struct-of-tag-and-unions"?

**Comment by poliorcetics** [comment]
> And finally, is it possible to summarize the difference between this and "normal" primitive reps in a sentence? Something like "union-of-structs vs struct-of-tag-and-unions"?

I'm not sure I understand what it is you want

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:443]:
```suggestion
For enums with fields, it is also possible to combine `repr(C)` and a
```

**Comment by RalfJung** [comment]
> > And finally, is it possible to summarize the difference between this and "normal" primitive reps in a sentence? Something like "union-of-structs vs struct-of-tag-and-unions"?
> 
> I'm not sure I understand what it is you want

I think it would be useful to directly contrast `repr(C)` and `repr(C, int)` and explain how they differ.

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:444]:
```suggestion
primitive representation (e.g., `repr(C, i32)`). This modifies the [`repr(C)`] by changing the
```

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md]:
```suggestion
representation of the discriminant enum to have the representation of the chosen
primitive. So, if you chose the `u8` representation, then the
```
At least I think that's what you mean? "the representation of the chose primitive representation" does make sense, I think.

**Comment by poliorcetics** [comment]
> I think it would be useful to directly contrast `repr(int)` and `repr(C, int)` and explain how they differ.

I added a new subsection that explains it and shows some example. 😄

File: src/type-layout.md
Comment (RalfJung) [reply src/type-layout.md]: This strange wording is still in the latest version I think.

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:447]: Here: "changing the representation [...] to have the representation of".  But representations do not have representations.

Comment (poliorcetics) [reply src/type-layout.md]: I updated the wording and added a little context to the example below, is it better this way ?

Comment (RalfJung) [reply src/type-layout.md]: Yes looks good. :)

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:380]: I'm a little unclear here.  "instead" of what?  `repr(C)`?

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:311]:
```suggestion
> convenient or consistent for them.
```

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:496]: This doesn't explain why two bytes would be equal to 8.

Also, it says `Enum8::Variant0`, but I would think it should be `EnumC::Variant0`?

Comment (poliorcetics) [reply src/type-layout.md:496]: That's a copy-paste mistake 😓

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md:380]: I'm not sure this whole note is useful. I could shorten it to:

> Note: Primitive representations are primarily intended for Rust code that wants to interoperate with the idioms of preexisting C/C++ codebases.

File: src/type-layout.md
Comment (ehuss) [inline src/type-layout.md:496]: I thought the discriminant was specified — as the size of a C `enum` of the target platform's C ABI?  Then there is 1 byte for the u8 value, plus padding.  It might also be worth indicating what platform this example is for, since the value could be different for other platforms?  Or maybe just scratch all that and say "the size is platform-dependent"?

Comment (poliorcetics) [reply src/type-layout.md:496]: I'll go with "the size is platform dependant"

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md]: I used the above wording in the "Fix nits 2" commit

Comment (ehuss) [reply src/type-layout.md]: Reading it again, I'm a little unclear, since there are multiple forms.  In this case, is it referring to `repr(C, Int)` or `repr(Int)`?  AIUI:

* `repr(C, Int)` is useful for C/C++ code that uses a struct with an explicitly-sized tag and payload (which is probably more common)
* `repr(Int)` is useful for C/C++ code that uses a union with an explicitly-sized tag in each struct, and can be more space-efficient than `repr(C, Int), but less common for pre-existing C code?
* `repr(C)` is useful for C/C++ code that uses a struct with a **C-enum tag** and payload? And that tag size is implementation-defined, and "best guess", so probably harder to get correct? That is, you really need to know which C compiler, platform, and C flags are being used?

Is that roughly correct?

Comment (poliorcetics) [reply src/type-layout.md]: I would say you are correct. I don't have much experience with this but the hobby projects I did used `repr(C, int)` to ensure maximal compatibility, which coincides with what you said. Would you like me to properly write in the document ?

Comment (ehuss) [reply src/type-layout.md]: Hm.  How about removing the note altogether for now?  This feels like it would be more useful in a guide on using FFI, and I don't want to dive too deep on it here. The reference has generally shied away from including information on how something is supposed to be used, as that tends more toward guide material.

**Comment by ehuss** [comment]
@RalfJung Did you have any further comments on this, or does it look good to you?

**Comment by poliorcetics** [comment]
Remind me to squash once all the changes are approved, I'm just waiting for all the reviews to be done. 😄

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md]: I removed it.

File: src/type-layout.md
Comment (RalfJung) [inline src/type-layout.md:314]: This seems to use the term "layout" for what is called "representation" above.

**Review by RalfJung — COMMENTED** [review]
My only remaining comment is that this seems to use "layout" and "representation" synonymously. But this is likely pre-existing and it is something we also struggle with in the UCG. (We mostly switched to "layout" there but then realized "layout" can mean at least 3 different things... see https://github.com/rust-lang/unsafe-code-guidelines/issues/122.)

I did not fact-check all the statements about representations and C++ here, as I am not deeply familiar with those details.

File: src/type-layout.md
Comment (poliorcetics) [reply src/type-layout.md:314]: No you're right, `representation` is used every where in those paragraphs, I'll change it.

**Review by ehuss — APPROVED** [review]
Thanks!  I appreciate sticking through with all the review comments.  I'll just squash via GitHub.

**Pull request merged**
